### PR TITLE
Add Annotation parsing and Config from Annotations functions

### LINKED ISSUE #330: Secrets Provider Configuration via annotations M1
Currently, Secrets Provider configuration is provided via environment variables. The goal of this issue is to make it possible to get this configuration from annotations.

## Context

The Secrets Provider container is not directly aware of the annotations from the manifest. The annotations are passed down via
the [downwards API](https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/#capabilities-of-the-downward-api). The format they take in file form is described as follows
> `metadata.annotations` - all of the pod's annotations, formatted as `annotation-key="escaped-annotation-value"` with one annotation per line

The code for the formatting is available [in the Kubernetes source](https://github.com/kubernetes/kubernetes/blob/0d9c29078b71fafe65261cb8597e6da0b1c8fe9e/pkg/fieldpath/fieldpath.go#L28-L41). This might be useful for parsing the annotations.

## Requirements

In order to support annotations as a mechanism for configuration we need to assume, as input, a file from the downward api containing the container's annotations. This files should be parsed and used to populate the same internal representations of configurations that were previously populated via environment variables.

Below is a mapping of internal configuration fields and source annotations.

Container config:
```markdown
PodNamespace - MY_POD_NAMESPACE (still comes from downward API)
RetryCountLimit - conjur.org/retry-count-limit
RetryIntervalSec - conjur.org/retry-interval-sec
StoreType - conjur.org/secrets-destination
```

Kubernetes Secrets config:
```markdown
RequiredK8sSecrets – conjur.org/k8s-secrets
```

The places in the code where this configuration is used are:

https://github.com/cyberark/secrets-provider-for-k8s/blob/53271bd8d884f7782df0c07e6619ec22568ecf50/pkg/secrets/k8s_secrets_storage/provide_conjur_secrets.go#L41

https://github.com/cyberark/secrets-provider-for-k8s/blob/53271bd8d884f7782df0c07e6619ec22568ecf50/cmd/secrets-provider/main.go#L31

## Acceptance Criteria
- [ ] An annotation parser exists for extracting arbitrary annotations from a downward API annotation file. This should include a test case to validate this behavior.
- [ ] Annotations can be used instead of environment variables. This should include a test case to validate this behavior.
- [ ] Annotations take precedence over environment variables. This should include a test case to validate this behavior.

## Notes on testing
For the test cases we recommend writing unit tests. E2E tests are very expensive to validate every aspect of configuration via annotations. It's better for the bulk of the tests to be unit tests, and as a finish touch to add annotations to a single E2E smoke test.

You can use the downwards API to generate static fixtures for the unit tests. Alternatively, the code for formatting is available [in the Kubernetes source](https://github.com/kubernetes/kubernetes/blob/0d9c29078b71fafe65261cb8597e6da0b1c8fe9e/pkg/fieldpath/fieldpath.go#L28-L41). Additionally, the formatting logic is available in a public package so you could also dynamically generate your fixtures.

```
package main

import (
	"fmt"

	"k8s.io/kubernetes/pkg/fieldpath"
)
func main() {
	fmt.Println(fieldpath.FormatMap(map[string]string{"moo": "meow"}))
}
```

### PR DESCRIPTION
**Pull request by john-odonnell**
### What does this PR do?

This PR allows for configuring Secrets Provider with Pod annotations. Annotation settings are prioritized over environment variable settings, but envvars can still be used to configure SP to maintain backwards compatibility.

#### Adds `annotations_parser.go` and `annotation_parser_test.go` to package `secrets/annotations`, including:
- `AnnotationsFromFile(path string) (map[string]string, error)`
- `ParseAnnotations(annotationsFile io.Reader) (map[string]string, error)`
- **Test Cases**
  - A valid annotations file is parsed successfully
  - Malformed annotations file lines result in an error indicating the line number and desired format

#### Adds functions to `secrets/config` package:
- `ValidateAnnotations(annotations map[string]string) ([]error, []error)`
  - Confirms annotation keys & value types, and confirms value for those annotations with enumerated value options
  - Returns lists of Error-level and Info-level logs
  - **Test Cases**
    - Properly formatted annotations return empty arrays
    - Confirm Info log if for unrecognized annotations with and without `conjur.org/` prefix
    - Confirm Error log for annotations given improper value type
    - Confirm Error log for annotations with explicit input ranges given invalid values
- `GatherSecretsProviderSettings(annotations map[string]string) map[string]string`
  - Returns an map containing annotations and envvars required from Secrets Provider config
  - **Test Cases**
    - If the annotations map is empty, the resulting map contains envvars used in SP config
    - If envvars are not set, the resulting map contains contains the annotations used in SP config
    - If both are present, the resulting map is all envvars and annotations used in SP config
    - Annotations and envvars not used in SP config are excluded
- `ValidateSecretsProviderSettings(envAndAnnots map[string]string) ([]error, []error)`
  - Confirm that the provided annotations and envvars lend to a valid SP config
  - Input validation on annotation keys and values
  - Returns Error and Info level log messages. An empty Error array means validation has succeeded.
  - **Test Cases**
    - Given a valid configuration of annotations, no errors are returned
    - Given a valid configuration of envvars, no errors are returned
    - Confirm that `MY_POD_NAMESPACE` is required in either case
    - Confirm Error log if StoreType is not set
    - Confirm Error log is Push-to-File is being configured with envvars
    - Confirm Error log if K8s Secrets mode is missing required K8s secrets
    - Confirm Info log if values are provided to equivalent annotation and envvar
- `NewConfig(settings map[string]string) *Config`
  - Create a new SP config from a validated map of annotations and SP Config envvars
  - **Test Cases**
    - A valid map of annotation-based settings results in a valid Config
    - A valid map of envvar-based settings results in a valid Config
    - A valid map of both annotation and envvar settings defers to annotation values

#### Acceptance Criteria Review
- [x] An annotation parser exists for extracting arbitrary annotations from a downward API annotation file. This should include a test case to validate this behavior.
- [x] Annotations can be used instead of environment variables. This should include a test case to validate this behavior.
- [x] Annotations take precedence over environment variables. This should include a test case to validate this behavior.
- [x] Unknown Conjur annotation causes INFO level log
- [x] Multiline annotation values are parsed correctly

### What ticket does this PR close?
ONYX-11827
Closes #330 

#### Change log
- [ ] The CHANGELOG has been updated, or
- [x] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [x] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [ ] This PR does not require updating any documentation

### CODE REVIEW

File: pkg/secrets/config/config_test.go [hunk 1]
```diff
@@ -1,76 +1,395 @@
...
+			StoreType:          "file",
+			RequiredK8sSecrets: []string{},
+			RetryCountLimit:    DEFAULT_RETRY_COUNT_LIMIT,
+			RetryIntervalSec:   DEFAULT_RETRY_INTERVAL_SEC,
+		}),
+	},
+}
+
+func TestValidateAnnotations(t *testing.T) {
+	for _, tc := range validateAnnotationsTestCases {
```
Comment (diverdane) [inline pkg/secrets/config/config_test.go:354]: This is COMPLETELY a personal preference sort of thing, but in the past, I've created table-driven tests where all of these are defined in one test function:

- The test case struct (this struct can be anonymous, if it's included in an instantiation of this struct, i.e. the test list)
- The test case list (e.g. `validateAnnotationsTestCases`)
- The iteration/running through all of the test cases

The benefit to this is that all definitions required for the test are gathered into one spot (one function). The downside is that the test function ends up with a LOT of lines (less visual breaks).

**Review by diverdane — COMMENTED** [review]
Really nice job weaving all of the different pieces of the configuration together (SP env variables, authn-k8s env variables, Annotations)! The resulting flow is pretty easy to follow, and it looks like you've covered all of the edge cases.

Nice work on the table-driven test cases and the use of closure for the assertion funcs!

I just have minor comments and some just-for-grins color commentary. Otherwise looks good-to-go.

File: pkg/secrets/config/config.go [hunk 2]
```diff
@@ -28,63 +29,266 @@ type Config struct {
...
+	// is not recognized as either a Secrets Provider config or Push to File config annotation
+	//
+	// If the annotation is for Push to File config, the ValidAnnotations function
+	// needs to be aware of the annotation's valid prefix in order to perform input validation,
+	// so this function returns:
+	//   - either the key, or the key's valid prefix
+	//   - the Map in which the key or prefix was found
+	//   - the success status of the operation
+	if !strings.HasPrefix(key, "conjur.org/") {
+		return "", nil, fmt.Errorf(messages.CSPFK011I, key)
```
[the remark this answers is not in the document]
Comment (john-odonnell) [reply pkg/secrets/config/config.go]: Yes, you're right. I think I initially included this case as a sanity check more than anything. I've updated the function to ignore annotations with the `conjur.org/` prefix.

This function, `validateAnnotationKey`, is called from `ValidateAnnotations` - at this point, it's assumed that the annotations map contains ALL annotations, regardless of validity.

File: pkg/secrets/annotations/annotation_parser.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/cyberark/conjur-authn-k8s-client/pkg/log"
+	"github.com/cyberark/secrets-provider-for-k8s/pkg/log/messages"
+)
+
+func AnnotationsFromFile(path string) (map[string]string, error) {
```
[the remark this answers is not in the document]
Comment (john-odonnell) [reply pkg/secrets/annotations/annotation_parser.go:17]: Noted, and gave all exported functions proper comments.
Is there a go-to linter for Go projects? I looked at [`golint`](https://github.com/golang/lint), but it looks like it's been deprecated for a while.

File: pkg/secrets/config/config.go [hunk 4]
```diff
@@ -28,63 +29,266 @@ type Config struct {
...
-		"MY_POD_NAMESPACE",
-		"K8S_SECRETS",
-		"SECRETS_DESTINATION",
-	} {
-		if os.Getenv(envvar) == "" {
-			return nil, log.RecordedError(messages.CSPFK004E, envvar)
+// Define supported annotation keys for Secrets Provider config, as well as value restraints for each
+var secretsProviderAnnotations = map[string][]string{
+	"conjur.org/authn-identity":      {"string"},
+	"conjur.org/container-mode":      {"string", "init", "application"},
```
[the remark this answers is not in the document]
Comment (john-odonnell) [reply pkg/secrets/config/config.go]: This approach is much more explicit, I like it!

File: pkg/secrets/config/config.go [hunk 5]
```diff
@@ -28,63 +29,266 @@ type Config struct {
...
+// Define environment variables used in Secrets Provider config
+var validEnvVars = []string{
+	"MY_POD_NAMESPACE",
+	"SECRETS_DESTINATION",
+	"K8S_SECRETS",
+	"RETRY_INTERVAL_SEC",
+	"RETRY_COUNT_LIMIT",
+}
+
+func ValidateAnnotations(annotations map[string]string) ([]error, []error) {
```
[the remark this answers is not in the document]
Comment (john-odonnell) [reply pkg/secrets/config/config.go:81]: Added comments to all exported functions in `config.go`.

File: pkg/secrets/config/config_test.go [hunk 6]
```diff
@@ -1,76 +1,395 @@
...
-
-			Convey("Invalid value of SECRETS_DESTINATION", func() {
-				_ = os.Setenv("SECRETS_DESTINATION", "invalid_store_type")
-
-				_, err := NewFromEnv()
-				Convey("Raises the proper error", func() {
-					So(err.Error(), ShouldEqual, fmt.Sprintf(messages.CSPFK005E, "SECRETS_DESTINATION"))
-				})
-			})
+func assertEmptyErrorList() func(*testing.T, []error, []error) {
```
[the remark this answers is not in the document]
Comment (john-odonnell) [reply pkg/secrets/config/config_test.go]: Done, called it `errorAssertFunc` to make it a bit more specific.

File: pkg/secrets/config/config.go [hunk 7]
```diff
@@ -28,63 +29,266 @@ type Config struct {
...
+	// validated that the value is of valid type, or confirm that the value is
+	// in the range of enumerated and acceptable values
+	switch targetType := acceptedValueInfo[0]; targetType {
+	case "int":
+		_, err := strconv.Atoi(value)
+		if err != nil {
+			return fmt.Errorf(messages.CSPFK042E, key, value, "Integer")
+		}
+	case "bool":
+		if value != "true" && value != "false" {
```
[the remark this answers is not in the document]
Comment (john-odonnell) [reply pkg/secrets/config/config.go:261]: No, not with this implementation. If we want, I can use `strconv.ParseBool` to determine if a boolean setting was acceptable, which would allow "TRUE" and "FALSE", and some others:

> ParseBool returns the boolean value represented by the string. It accepts 1, t, T, TRUE, true, True, 0, f, F, FALSE, false, False. Any other value returns an error.

File: cmd/secrets-provider/main.go [hunk 8]
```diff
@@ -75,14 +100,14 @@ func main() {
 	}
 }
 
-func provideSecretsToTarget(authn *authenticator.Authenticator, provideConjurSecrets secrets.ProvideConjurSecrets, accessToken *memory.AccessToken) error {
+func provideSecretsToTarget(authn *authenticator.Authenticator, provideConjurSecrets secrets.ProvideConjurSecrets, accessToken *memory.AccessToken, secretsConfig *secretsConfigProvider.Config) error {
```
[the remark this answers is not in the document]
Comment (john-odonnell) [reply cmd/secrets-provider/main.go]: Wrapped arguments, two per line.

File: pkg/secrets/annotations/annotation_parser.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (diverdane) [reply pkg/secrets/annotations/annotation_parser.go:17]: Looks like `revive` is pretty standard, and is comparable to what `golint` was.
I installed and ran `revive` manually with:
```
go install github.com/mgechev/revive@latest
revive -formatter friendly
```

When I removed the doc comment that you just added and ran `revive`, I see:
```
$ revive -formatter friendly
  ⚠  https://revive.run/r#exported  exported function AnnotationsFromFile should have comment or be unexported  
  annotation_parser.go:14:1

  ⚠  https://revive.run/r#exported  func name will be used as annotations.AnnotationsFromFile by other packages, and that stutters; consider calling this FromFile  
  annotation_parser.go:14:6

⚠ 2 problems (0 errors, 2 warnings)

Warnings:
  2  exported  

$
```

I'll have to see if I can add this to my Go-Vim config (and/or Goland).

[the remark this answers is not in the document]
Comment (john-odonnell) [reply pkg/secrets/annotations/annotation_parser.go]: Ran `revive` on the files I changed, and had to update some constants in `config.go` that propagated to a couple other files. Had to convert from `THIS_CASE` to `ThisCase`.

**Review by diverdane — APPROVED** [review]
LGTM!!!
